### ISSUE TITLE
bug: using package exports fields can result in incompatible types for identical types

### ISSUE THREAD

**Issue by EisenbergEffect** [microsoft/TypeScript#49466]
# Bug Report

### 🔎 Search Terms

* incompatible type

### 🕗 Version & Regression Information

This bug pertains to the new support for Node16 package exports.

- This is an unexpected compilation error.
- This occurs in TypeScript 4.7.3.
- I was unable to test this on prior versions because it pertains to a feature introduced in 4.7.

### ⏯ Playground Link

A playground link will not demonstrate the issue because it only occurs when importing a second package's exports.

### 💻 Code

This PR demonstrates a set of code changes to the `@microsoft/fast-element` library which results in a package that cannot be properly consumed from TypeScript: https://github.com/microsoft/fast/pull/6087

The PR simply moves some code, which was originally part of the default export, and makes it an optional set of functionality exposed only through a special package export via `@microsoft/fast-element/binding/two-way`. 

This is an important scenario for us and a dependency of our upcoming release.

### 🙁 Actual behavior

As a test, I built the package, which builds fine. Then I created a new project where I could consume the package. Since we haven't published this package yet (or even merged the PR), I made a quick check by simply copying my local version of the package over into the `node_modules` folder of my test project. I then attempted to import via the path above. There is no problem importing. The problem is that the types between the default export and the special export don't match, according to the compiler. So, if you have a function that takes a certain type as an input, coming from the default export, and then the special export exports an implementor of that interface, you cannot pass the instance to the function. Here is a made up example of what I'm talking about:

#### `@foo/bar` `index.ts`

```ts
export interface Task {
  run();
}

export function runTask(task: Task) {
  task.run();
}
```

#### `@foo/bar` `cool-tasks.ts`

```ts
import { Task } from "./index.js";

export const coolTask: Task = {
  run() {
    // ellided
  }
};
```

#### `@foo/bar` `package.json`

```json
"exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "production": "./dist/esm/index.js",
      "development": "./dist/esm/index.debug.js",
      "default": "./dist/esm/index.js"
    },
    "./cool-tasks": {
      "types": "./dist/dts/cool-tasks.d.ts",
      "default": "./dist/esm/cool-tasks.js"
    }
  }
```

#### Usage in a Project

```ts
import { runTask } from "@foo/bar";
import { coolTask } from "@foo/bar/cool-tasks";

runTask(coolTask); // Compiler error!!! Types are incompatible!
```

Note: I have not tried out this exact code, but this is essentially the pattern from our package with the PR linked above. We cannot pass the `twoWay` `BindingConfig` const to the `bind` function that takes an instance `BindingConfig`. TypeScript thinks that the types are incompatible, even though they are both referencing the same file in the package.

Here is a screenshot of the real code in action:

![MicrosoftTeams-image](https://user-images.githubusercontent.com/131485/173089693-1d48c54b-321a-4f0d-8948-7ab42974efa9.png)

This *exact* same consuming code works perfectly if I simply move the `twoWay` code back into the default export and import it from there. If you examine the PR above, you will see that all we have done is moved the code to its own file and specified a package export.

### 🙂 Expected behavior

I would expect this to compile without issue.

**Comment by EisenbergEffect** [comment]
Ok, I have discovered a way to easily reproduce the error with code we have already shipped in beta. Here are the instructions:

* `npm install --save @microsoft/fast-element@2.0.0-beta.1`
* Attempt to compile the code below

```ts
import { ArrayObserver, Observable } from "@microsoft/fast-element";
import { mergeSpliceStrategy } from "@microsoft/fast-element/splice-strategies";

const observer = Observable.getNotifier<ArrayObserver>([]);
observer.strategy = mergeSpliceStrategy;
```

You will see the same type of error as above when attempting to set the `strategy` property. That property is of type `SpliceStrategy` and the `strategy` property is also of type `SpliceStrategy`. The package source internally imports this interface from the same file. However, when the strategy implementation is imported from a package export, TypeScript believes that the types are incompatible. If I move the strategy implementation code into the main package export, there is no error.

This is a major ship-blocker for FAST 2.0. We really appreciate your assistance in resolving this issue. Please let me know if there's anything I can do to help.

**Comment by weswigham** [comment]
This is because you have two, distinct copies of `SpliceStrategy` - one in `./dist/fast-element.d.ts` (the one reachable via the types in the root import), and one in `./dist/dts/observation/array.d.ts` (thje one reachable via the deep import). Since it has a private, it is nominal, thus the two instances of the similarly named classes are incompatible with one another.

If you have deep entrypoints, you shouldn't be bundling you `.d.ts` file for your root entrypoint.

**Comment by EisenbergEffect** [comment]
@weswigham Are you saying that if we point our root d.ts export to the non-bundled index.d.ts then the problems with just "magically" disappear? In other words, do we just change this:

```json
".": {
      "types": "./dist/fast-element.d.ts",
      "production": "./dist/esm/index.js",
      "development": "./dist/esm/index.debug.js",
      "default": "./dist/esm/index.js"
    },
```

Into this:

```json
".": {
      "types": "./dist/index.d.ts", <-- just change this?
      "production": "./dist/esm/index.js",
      "development": "./dist/esm/index.debug.js",
      "default": "./dist/esm/index.js"
    },
```

**Comment by EisenbergEffect** [comment]
Thanks @weswigham! I'll give this a try. It makes sense. I'll report back soon.

**Comment by EisenbergEffect** [comment]
A quick local test (updating some package.json in node modules folder) shows that this indeed gets it working beautifully. Again, huge thanks @weswigham !!! Closing issues as resolved.

**Comment by EisenbergEffect** [comment]
@weswigham QQ: If I'm using exports as above, should I also remove the "types" package.json filed that previous also pointed to the rollup dts? Am I right in assuming that with exports configuration that the "types" field is no longer needed?

**Comment by weswigham** [comment]
Keep the `types` field for people running older versions of node/ts, but you're free to point it at the unbundled entry point, too.

**Comment by EisenbergEffect** [comment]
Thank you!
